core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -512,7 +546,58 @@ impl<Block: BlockT> Inner<Block> {
 		self.multicast_neighbor_packet()
 	}
 
-	fn consider_vote(&self, round: Round, set_id: SetId) -> Consider {
+	fn consider_vote(&mut self, who: &PeerId, round: Round, set_id: SetId, msg: &SignedMessage<Block>) -> Consider {
+		// A tally of messages received per peer, per voter, per round.
+		let peer_tally_for_round = match self.incoming_msg_tally.get_mut(&(round, set_id)) {
+			Some(tally_for_round) =>  tally_for_round,
+			None => {
+				// We don't know about this round,
+				// let the local-view handle it and likely treat it as past or future.
+				return self.local_view.as_ref().map(|v| v.consider_vote(round, set_id))
+					.unwrap_or(Consider::RejectOutOfScope)
+			}
+		};
+		let mut per_peer_tally = peer_tally_for_round.entry((msg.id.clone(), who.clone())).or_insert(Default::default());
+
+		// A tally of messages we sent out, per peer, per voter, per round
+		let outgoing_tally_for_round = self.outgoing_msg_tally
+			.get_mut(&(round, set_id))
+			.expect("If incoming votes knows about this round, so should outgoing ones; qed");
+		let outgoing_tally = outgoing_tally_for_round.entry((msg.id.clone(), who.clone())).or_insert(Default::default());
+
+		match &msg.message {
+			PrimaryPropose(_propose) => {
```

Comment:
I think right now we allow multiple proposals as long as they are from the primary. Not sure if we should have this strict check here. Honest primaries should only propose once though. cc @rphmeier

Reply:
the code committed appears to be right